rfcs/003-strings.md

```diff
@@ -0,0 +1,325 @@
+# Strings in windows-rs
+
+## Summary
+
+Rust and Windows each have a fair amount of string types. This RFC proposes a system for making it easy and obvious how to make the string types from both environments work well together.
+
+## Motivation
+
+Making Rust and Windows string types work together is integral to ensuring correct, performant, and boiler-plate free code.
+
+### What we want to accomplish
+
+In general, we are trying to accomplish two goals:
+
+* Correctly model the Windows string types in a way that leverages Rust's type and ownership systems to prevent incorrect use.
+* Allow for easy and low cost ways to convert between Windows and Rust types.
+
+On top of this, we want these APIs to feel natural and obvious to Rust developers.
+
+## Explanation
+
+### The Different Types of Strings
+
+#### Windows
+
+Windows has the following string types:
+
+* `PSTR`: a mutable, nul-terminated string (i.e., composed of "characters" a.k.a. `u8`s) that is often but not always ANSI encoded.
+* `PCSTR`: an immutable version of `PSTR`.
+  * **QUESTION**: what are the practical differences between this and `PSTR`?
+* `PWSTR`: a mutable, nul-terminated "wide" string (i.e., composed of "wide characters" a.k.a. `u16`s) that is often but not always UTF-16 encoded.
```

Comment:
> [...] that is often but not always UTF-16 encoded.

Can you expand on this? I'm not aware of any non-UTF-16 encoded PWSTRs. Are you instead trying to convey that a grapheme may span 1 or 2 code 16-bit units (2-4 bytes)?

Reply:
It's possible to create [invalid UTF-16 strings](https://www.unicode.org/faq/utf_bom.html#utf16-7) because the kernel does not check validity. This is exceedingly rare in practice but it is possible, especially if there's a malicious app causing mischief.

Reply:
The prime candidate here is the filesystem. NTFS allows names to be just about *any* sequence of 16-bit values (with a few reserved code units). Those names will inevitably be returned from `FindFirstFile`/`FindNextFile`, so `PWSTR`/`PWCSTR` cannot make any assumptions about the validity of the UTF-16 encoding.

Likewise, named kernel objects (events, mutexes, pipes, ...) can have names that aren't valid UTF-16.

Not quite the norm, but invalid UTF-16 sequences are observed in the wild, usually as a result of an accident/bug, or, more frequently, in attacks.

Reply:
There's no use trying to define what these are. They're just pointer `typedef`s, which is why `windows-rs` defines them as simply as possible. We can provide safe ways to create them and pass them to API calls, but we cannot make any guarantees about them or provide generally safe conversions from them.